File: ai-project_front_end/src/lib/api/uiAutomation.ts
import { requestJson, authHeader } from './client'

export interface UiTestScript {
  id: string
  projectId: string
  name: string
  description: string
  scriptType: string
  scriptContent: string
  recordingJson: Record<string, unknown>
  status: string
  browser: string
  viewportWidth: number
  viewportHeight: number
  baseUrl: string
  tags: string[]
  createdBy: string | null
  createdAt: number
  updatedAt: number
}

export interface UiTestRun {
  id: string
  scriptId: string
  status: string
  startedAt: string | null
  finishedAt: string | null
  durationMs: number | null
  stepsTotal: number
  stepsPassed: number
  stepsFailed: number
  screenshotPaths: string[]
  errorMessage: string | null
  tracePath: string | null
  reportPath: string | null
  createdAt: number
  updatedAt: number
}

export interface PageData<T> {
  page: number
  pageSize: number
  total: number
  items: T[]
}

export async function listScripts(projectId: string, page = 1, pageSize = 20): Promise<PageData<UiTestScript>> {
  return requestJson<PageData<UiTestScript>>(
    `/api/projects/${projectId}/ui-automation/scripts?page=${page}&pageSize=${pageSize}`,
    { method: 'GET', headers: authHeader() }
  )
}

export async function getScript(projectId: string, scriptId: string): Promise<UiTestScript> {
  return requestJson<UiTestScript>(
    `/api/projects/${projectId}/ui-automation/scripts/${scriptId}`,
    { method: 'GET', headers: authHeader() }
  )
}

export async function createScript(projectId: string, data: {
  name: string; description?: string; scriptType?: string;
  browser?: string; viewportWidth?: number; viewportHeight?: number;
  baseUrl?: string; tags?: string[];
}): Promise<UiTestScript> {
  return requestJson<UiTestScript>(
    `/api/projects/${projectId}/ui-automation/scripts`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...authHeader() },
      body: JSON.stringify(data),
    }
  )
}

export async function updateScript(projectId: string, scriptId: string, data: Partial<UiTestScript>): Promise<UiTestScript> {
  return requestJson<UiTestScript>(
    `/api/projects/${projectId}/ui-automation/scripts/${scriptId}`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json', ...authHeader() },
      body: JSON.stringify(data),
    }
  )
}

export async function deleteScript(projectId: string, scriptId: string): Promise<void> {
  await requestJson<void>(
    `/api/projects/${projectId}/ui-automation/scripts/${scriptId}`,
    { method: 'DELETE', headers: authHeader() }
  )
}

export async function saveRecording(projectId: string, scriptId: string, actions: Record<string, unknown>[]): Promise<UiTestScript> {
  return requestJson<UiTestScript>(
    `/api/projects/${projectId}/ui-automation/scripts/${scriptId}/record`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...authHeader() },
      body: JSON.stringify({ actions }),
    }
  )
}

export async function generateScript(projectId: string, scriptId: string): Promise<UiTestScript> {
  return requestJson<UiTestScript>(
    `/api/projects/${projectId}/ui-automation/scripts/${scriptId}/generate`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...authHeader() },
      body: JSON.stringify({}),
    }
  )
}

export async function runTest(projectId: string, scriptId: string): Promise<UiTestRun> {
  return requestJson<UiTestRun>(
    `/api/projects/${projectId}/ui-automation/scripts/${scriptId}/run`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', ...authHeader() },
      body: JSON.stringify({}),
    }
  )
}

export async function listRuns(projectId: string, scriptId?: string, page = 1, pageSize = 20): Promise<PageData<UiTestRun>> {
  const params = new URLSearchParams({ page: String(page), pageSize: String(pageSize) })
  if (scriptId) params.set('scriptId', scriptId)
  return requestJson<PageData<UiTestRun>>(
    `/api/projects/${projectId}/ui-automation/runs?${params}`,
    { method: 'GET', headers: authHeader() }
  )
}

export async function getRun(projectId: string, runId: string): Promise<UiTestRun> {
  return requestJson<UiTestRun>(
    `/api/projects/${projectId}/ui-automation/runs/${runId}`,
    { method: 'GET', headers: authHeader() }
  )
}

File: ai-project-back-end/app/schemas/ui_automation.py
from __future__ import annotations

from typing import Any

from pydantic import Field

from app.schemas.common import BaseSchema


class UiTestScriptCreateRequest(BaseSchema):
    name: str = Field(min_length=1, max_length=255)
    description: str = ""
    projectId: str = Field(min_length=1, max_length=64)
    scriptType: str = Field(default="PLAYWRIGHT", max_length=32)
    browser: str = Field(default="chromium", max_length=32)
    viewportWidth: int = Field(default=1280, ge=320, le=3840)
    viewportHeight: int = Field(default=720, ge=240, le=2160)
    baseUrl: str = Field(default="", max_length=512)
    tags: list[str] = Field(default_factory=list)


class UiTestScriptUpdateRequest(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    scriptType: str | None = Field(default=None, max_length=32)
    browser: str | None = Field(default=None, max_length=32)
    viewportWidth: int | None = Field(default=None, ge=320, le=3840)
    viewportHeight: int | None = Field(default=None, ge=240, le=2160)
    baseUrl: str | None = Field(default=None, max_length=512)
    tags: list[str] | None = None


class UiTestScriptDetail(BaseSchema):
    id: str
    projectId: str
    name: str
    description: str = ""
    scriptType: str = "PLAYWRIGHT"
    scriptContent: str = ""
    recordingJson: dict[str, Any] = Field(default_factory=dict)
    status: str = "DRAFT"
    browser: str = "chromium"
    viewportWidth: int = 1280
    viewportHeight: int = 720
    baseUrl: str = ""
    tags: list[str] = Field(default_factory=list)
    createdBy: str | None = None
    createdAt: int = 0
    updatedAt: int = 0


class UiTestRunDetail(BaseSchema):
    id: str
    scriptId: str
    status: str = "QUEUED"
    startedAt: str | None = None
    finishedAt: str | None = None
    durationMs: int | None = None
    stepsTotal: int = 0
    stepsPassed: int = 0
    stepsFailed: int = 0
    screenshotPaths: list[str] = Field(default_factory=list)
    errorMessage: str | None = None
    tracePath: str | None = None
    reportPath: str | None = None
    createdAt: int = 0
    updatedAt: int = 0


class UiTestRecordRequest(BaseSchema):
    actions: list[dict[str, Any]] = Field(min_length=1)


class UiTestGenerateRequest(BaseSchema):
    pass


class UiTestRunRequest(BaseSchema):
    pass
